File: backend/SaidaPessoas.API/Controllers/UsuariosController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System.Security.Claims;
using SaidaPessoas.API.Data;
using SaidaPessoas.API.DTOs;

namespace SaidaPessoas.API.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class UsuariosController : ControllerBase
{
    private readonly AppDbContext _context;

    public UsuariosController(AppDbContext context) => _context = context;

    private int GetUserId() => int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)!.Value);

    [HttpGet("me")]
    public async Task<IActionResult> GetMe()
    {
        var usuario = await _context.Usuarios.FindAsync(GetUserId());
        if (usuario == null) return NotFound();

        return Ok(new UsuarioResponseDto(
            usuario.Id, usuario.Nome, usuario.Matricula, usuario.Email,
            usuario.Setor, usuario.Perfil.ToString(), usuario.Status.ToString(),
            usuario.DataCadastro
        ));
    }

    [HttpPut("me/senha")]
    public async Task<IActionResult> AlterarMinhaSenha([FromBody] AlterarSenhaPropriaDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.NovaSenha) || dto.NovaSenha.Length < 6)
            return BadRequest(new { message = "A nova senha deve ter no mínimo 6 caracteres." });

        var usuario = await _context.Usuarios.FindAsync(GetUserId());
        if (usuario == null) return NotFound();

        if (!BCrypt.Net.BCrypt.Verify(dto.SenhaAtual, usuario.SenhaHash))
            return BadRequest(new { message = "Senha atual incorreta." });

        usuario.SenhaHash = BCrypt.Net.BCrypt.HashPassword(dto.NovaSenha);
        await _context.SaveChangesAsync();
        return Ok(new { message = "Senha alterada com sucesso." });
    }
}

File: backend-node/src/usuarios/usuarios.controller.ts
import { Body, Controller, Get, NotFoundException, BadRequestException, Put } from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import { PrismaService } from '../prisma/prisma.service';
import { AuthUser, CurrentUser } from '../common/decorators';
import { toInstant } from '../common/dates';

class AlterarSenhaPropriaDto {
  senhaAtual!: string;
  novaSenha!: string;
}

@Controller('usuarios')
export class UsuariosController {
  constructor(private readonly prisma: PrismaService) {}

  @Get('me')
  async getMe(@CurrentUser() user: AuthUser) {
    const u = await this.prisma.usuarios.findUnique({ where: { Id: user.userId } });
    if (!u) throw new NotFoundException();
    return {
      id: u.Id,
      nome: u.Nome,
      matricula: u.Matricula,
      email: u.Email,
      setor: u.Setor,
      perfil: u.Perfil,
      status: u.Status,
      dataCadastro: toInstant(u.DataCadastro),
    };
  }

  /** Colaboradores ativos — usado no formulário de saída por terceiros. */
  @Get('colaboradores')
  async listarColaboradores() {
    const usuarios = await this.prisma.usuarios.findMany({
      where: { Status: 'Ativo' },
      select: { Id: true, Nome: true, Setor: true, Matricula: true },
      orderBy: { Nome: 'asc' },
    });
    return usuarios.map((u) => ({ id: u.Id, nome: u.Nome, setor: u.Setor, matricula: u.Matricula }));
  }

  @Put('me/senha')
  async alterarMinhaSenha(@CurrentUser() user: AuthUser, @Body() dto: AlterarSenhaPropriaDto) {
    if (!dto.novaSenha || dto.novaSenha.length < 6) {
      throw new BadRequestException({ message: 'A nova senha deve ter no mínimo 6 caracteres.' });
    }
    const u = await this.prisma.usuarios.findUnique({ where: { Id: user.userId } });
    if (!u) throw new NotFoundException();
    if (!bcrypt.compareSync(dto.senhaAtual, u.SenhaHash)) {
      throw new BadRequestException({ message: 'Senha atual incorreta.' });
    }
    await this.prisma.usuarios.update({
      where: { Id: user.userId },
      data: { SenhaHash: bcrypt.hashSync(dto.novaSenha, 10) },
    });
    return { message: 'Senha alterada com sucesso.' };
  }
}
